tools/kagami/src/genesis.rs

```diff
@@ -241,10 +259,231 @@ fn generate_synthetic(
                     ),
                 )
                 .into();
-                first_transaction.append_instruction(mint);
+                first_transaction.push_instruction(mint);
             }
         }
     }
 
     Ok(genesis)
 }
+
+/// Use `Kagami` to sign genesis block.
+#[derive(Clone, Debug, Parser)]
+#[command(group = ArgGroup::new("private_key").required(true))]
+#[command(group = ArgGroup::new("public_key").required(true))]
+pub struct SignArgs {
```

Comment:
This should not be in this module. We want facility to sign all transactions, not just genesis. Genesis is only a special case

Reply:
@0x009922 do we put it in `kagami crypto` or will we have another command?

Reply:
now that I think about it the command should be like this `kagami crypto sign file`. It can sign any file. Contents of the file should not be relevant. Whether it's transaction or genesis transaction or query

Reply:
At the moment I rely on the fact that the file is indeed genesis file - there are verifications for executor via `RawGenesisBlock` struct. 
What if I move that logic to crypto and leave it somewhat "as is" and we create a task to extend the functionality somewhere down the road? We might never need it and then we just abandon the idea.
Will it be fine?

Reply:
@mversic, I proposed to move it to `kagami genesis` from `kagami crypto` deliberately.

It's not just signing of a file. In case of `genesis.json`, Kagami needs to know that it is a genesis file, to decoded it from JSON properly, read `executor_file`, combine it all into a single genesis transaction, and then produce a signature of it. It is quite different from signing any file, interpreting it as just some blob of bytes.

Therefore, I still think `kagami genesis` is a proper place for it, since it is tightly coupled to the genesis.

Reply:
am I missing something. When did we deviate from [this](https://github.com/hyperledger/iroha/issues/4225#issuecomment-1905464604)? Genesis file should exactly contain one JSON serialized `SignedTransaction`. We don't want no special genesis transaction type if we can avoid it

To deserialize this JSON file, `kagami` only needs to know that it's dealing with any `SignedTransaction` not that it's genesis. So it can work for any transaction. To make it work for multiple different types we `kagami` can try deserializing into multiple types `SignedTransaction`, `SignedQuery` (is there any other?) or it can take a CLI argument such as `--type query`

Reply:
Yes.
Here, you approved @0x009922's [idea](https://github.com/hyperledger/iroha/pull/4382/#discussion_r1540642426)

Reply:
Yes, but I don't understand how that is related to the discussion here

Reply:
Let me know if I don't understand you.
You suggest to sign genesis offline and keep it as a `SignedTransaction`. But previously we agreed on separating "signature" and genesis file itself as a separate entities.

Reply:
Just leave the signatures field empty in the file. It's a list so it can be empty

Reply:
> am I missing something. When did we deviate from https://github.com/hyperledger/iroha/issues/4225#issuecomment-1905464604? Genesis file should exactly contain one JSON serialized SignedTransaction. We don't want no special genesis transaction type if we can avoid it

Indeed, we deviated. I pictured the solution differently.

So, @mversic, you propose to make `genesis.json` to be a direct representation of `SignedTransaction`. Which means, it will contain a complete payload and its signature(s) in a final format. It won't have a convenience such as `"executor_file": "./executor.wasm"`.

If so, than I am confused, as it is going to increase complexity of what users need to do to prepare the genesis. They will need to prepare some unsigned transaction, put an "Upgrade" with executor into it, then sign it. Along the way they should also put chain id somewhere in the right place (even I don't know where exactly). The resulting `genesis.json` will also be bloated due to the inlined executor binary.

In the meantime, we can:

- keep current "convenient" `genesis.json` (with an addition of `chain_id` parameter)
- specify it by `genesis.file` config
- produce its signature by `kagami genesis sign genesis.json`
- put the signature into `genesis.signature` config

I might be missing something.

Reply:
Ok. The point about executor makes sense. Can someone write an example of how genesis will look like?

Reply:
I expect it to remain essentially the same. Only with addition of `chain_id`, and changing of `transactions: Instruction[][]` to `transaction: Instruction[]`

```json
{
  "chain_id": "0-0",
  "transaction": [
    {
      "Register": {
        "Domain": {
          "id": "wonderland",
          "logo": null,
          "metadata": {
            "key": {
              "String": "value"
            }
          }
        }
      }
    }
  ],
  "executor_file": "./executor.wasm"
}
```

Reply:
before I comment on this tell me this. Do you think genesis should be given with `--submit-genesis` or should it be submitted like any other transaction with `iroha_client`? We no longer need to submit genesis at the time of bootstrapping the network. Admin can bootstrap the network and push genesis later to 1 or multiple nodes

Reply:
let me disambiguate this. We have:
1. what `kagami sign` takes. This can have any JSON schema
2. what `--submit-genesis` takes. This should have the JSON schema of `SignedTransaction`

As far as I'm concerned 1. can be whatever. But 2. must be `SignedTransaction` or at best `TransactionPayload`. Nothing else should be omitted because the node needs all the info to verify the signature. If you omit a field (such as `time_to_live_ms`) then we have to, for each omitted field, either:
1. provide it as another argument together with `--submit-genesis`
2. we have to agree that every genesis will have some predefined value for this field

option 1. I will not even entertain. Option 2. is too implicit for my taste. This tells me that `kagami sign` should best produce full genesis, not just a signature. And this produced genesis should be SCALE encoded blob because it will have inlined executor or some other smart contract

Reply:
By specifying `time_to_live_ms` do you talk about signed genesis or just any block in general?
One of the issues for us to have a full `SignedTransaction` as a parameter is its size - right now even SCALE encoded version is around 1MB.
Also, before that we didn't have a parameter for `--signed-genesis`, it was a flag. Do you want to change that?

Reply:
I've lost thread of what we are trying to achieve...

> Do you think genesis should be given with --submit-genesis or should it be submitted like any other transaction with iroha_client?

If there is no technical need to have the genesis block anymore, then why have it? Let's get rid of it.